src/core/components/pin.js

```diff
@@ -0,0 +1,392 @@
+        }
+        if (type === types.indirect || type === types.all) {
+          getIndirectKeys((err, indirects) => {
+            if (err) { return callback(err) }
+            pins = pins
+              // if something is pinned both directly and indirectly,
+              // report the indirect entry
+              .filter(({ hash }) =>
+                !indirects.includes(hash) ||
+                (indirects.includes(hash) && !directPins.has(hash))
+              )
+              .concat(indirects.map(hash => ({
+                type: types.indirect,
+                hash
+              })))
+            return callback(null, pins)
+          })
+        } else {
+          return callback(null, pins)
+        }
+      }
+    }),
+
+    _isPinnedWithType: promisify((multihash, type, callback) => {
+      const key = toB58String(multihash)
+      const { recursive, direct, all } = types
+      // recursive
+      if ((type === recursive || type === all) && recursivePins.has(key)) {
+        return callback(null, {pinned: true, reason: recursive})
+      }
+      if ((type === recursive)) {
+        return callback(null, {pinned: false})
+      }
+      // direct
+      if ((type === direct || type === all) && directPins.has(key)) {
+        return callback(null, {pinned: true, reason: direct})
+      }
+      if ((type === direct)) {
+        return callback(null, {pinned: false})
+      }
+
+      // indirect (default)
+      // check each recursive key to see if multihash is under it
+      // arbitrary limit, enables handling 1000s of pins.
+      let foundPin
+      async.someLimit(recursiveKeys(), concurrencyLimit, (key, cb) => {
+        dag.get(new CID(key), (err, res) => {
+          if (err) { return cb(err) }
+
+          pinset.hasDescendant(res.value, multihash, (err, has) => {
+            if (has) {
+              foundPin = toB58String(res.value.multihash)
+            }
+            cb(err, has)
+          })
+        })
+      }, (err, found) => {
+        if (err) { return callback(err) }
+        return callback(null, { pinned: found, reason: foundPin })
+      })
+    }),
+
+    _load: promisify(callback => {
+      async.waterfall([
+        // hack for CLI tests
+        (cb) => repo.closed ? repo.datastore.open(cb) : cb(null, null),
+        (_, cb) => repo.datastore.has(pinDataStoreKey, cb),
+        (has, cb) => has ? cb() : cb(new Error('No pins to load')),
+        (cb) => repo.datastore.get(pinDataStoreKey, cb),
+        (mh, cb) => dag.get(new CID(mh), cb)
+      ], (err, pinRoot) => {
+        if (err) {
+          if (err.message === 'No pins to load') {
+            self.log('No pins to load')
+            return callback()
+          } else {
+            return callback(err)
+          }
+        }
+
+        async.parallel([
+          cb => pinset.loadSet(pinRoot.value, types.recursive, cb),
+          cb => pinset.loadSet(pinRoot.value, types.direct, cb)
+        ], (err, [rKeys, dKeys]) => {
```

Comment:
You can't destructure like this in callbacks, if an error occurred then the result will be null/undefined this will throw.

Reply:
That is so rad 💯

Reply:
Yeaass! :D